design/LoadBalancing/LoadBalancing.md

```diff
@@ -0,0 +1,221 @@
+
+* Commit proxy interface
+* GetReadVersion proxy interface
+* ConfigFollower interface
+
+The interface is assumed to be always *fresh*, i.e. the list of the servers is fixed.
+
+```mermaid
+graph LR
+    H0{Has alternatives?}
+    H1[Pick up an alternative]
+    H2[Backoff]
+    H3[Request]
+    H4([Reply])
+    H5([Error])
+    H6([Never])
+    H((Start)) --> H0
+    H0 --No--> H6
+    H0 --Yes--> H1
+    H1 --No healthy alternatives--> H2 --> H1
+    H1 --Has alternative--> H3 --Success--> H4
+    H3 --Exception--> H5
+    H3 --Broken Promise --> H2
+```
+
+### Alternative pick up algorithm
+
+In `basicLoadBalance`, a *best* alternative is picked up and used at the beginning. At this stage, this alternative is randomly picked up among all alternatives. If the best alternative does not work, it will iteratively try other interfaces, see [here](#picking-up-an-alternative-in-basic-load-balancing-algorithm).
+
+## `loadBalance`
+
+`loadBalance` provides a more sophisticated implementation of load balancing. In addition of the basic load balancing, it also provides a variety of features, such like
+
+* Support for Test Storage Server ([TSS](https://github.com/apple/foundationdb/blob/main/documentation/sphinx/source/tss.rst))
+* Distance-based candidate election
+* Able to handle timeouts and exceptions with retries
+* etc.
+
+Currently it is used for
+
+* Storage Server interface
+* BlobWorker interface
+
+
+
+```mermaid
+graph LR
+    H((Start))
+    H0{Has alternatives?}
+    H1[Choose initial candidates]
+    H4([Never])
+    H5[Pick up an alternative]
+    H6[Send request]
+    H7[Wait for available alternative]
+    H8([Response])
+    H9([All alternatives failed])
+
+    H --> H0 --No--> H4
+    H0 --Yes--> H1
+    H1 --> H5
+    H5 --Has alternative--> H6
+    H5 --No alternative-->H7
+    H6 --Success--> H8
+    H6 --Failure--> H5
+    H7 --At least one alternative--> H5
+    H7 --> H9
+```
+
+Note:
+
+* Response could be an exception, e.g. `process_behind` or `request_maybe_delivered`, and will be delivered as `Error` to the caller.
+
+### Choose initial candidates
+
+Two initial candidates will be picked up before the requests start. They will be selected as the first two alternatives for the load balancer. If both of them failed, other alternatives are used in a round-robin way.
+
+#### No `QueueModel`
+
+If no `QueueModel` is provided, the initial candidates are picked up randomly. The first candidate, or the *best* alternative, will always be one of local workers.
+
+#### With `QueueModel`
+
+`QueueModel` holds information about each candidate related to future version, latency and penalty.
+
+* If the storage server is returning a future version error, it is marked as not available until some certain time.
+* Penalty is reported by storage server in each response (see `storageserver.actor.cpp:StorageServer::getPenalty`). It is determined by the write queue length and the version lagging.
```

Comment:
`version lagging` is typically used to indicate the lag between tlog and the storage server, but here, the lag the version lag between SS latest version and the SS durable version. Consider changing `version lagging` to `durability lag`.

Reply:
Good to know this difference.